multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/CheckServicesToDeleteStep.java

```diff
@@ -15,12 +25,42 @@
 @Scope(BeanDefinition.SCOPE_PROTOTYPE)
 public class CheckServicesToDeleteStep extends CheckForOperationsInProgressStep {
 
+    @Inject
+    private ApplicationConfiguration applicationConfiguration;
+
     @Override
-    protected List<CloudServiceInstanceExtended> getServicesToProcess(ProcessContext context) {
+    protected Set<CloudServiceInstanceExtended> getExistingServicesToProcess(ProcessContext context) {
         List<String> servicesToDelete = context.getVariable(Variables.SERVICES_TO_DELETE);
-        return servicesToDelete.stream()
-                               .map(this::buildCloudServiceExtended)
-                               .collect(Collectors.toList());
+        CloudControllerClient client = context.getControllerClient();
+
+        int maxParallelThreads = Math.min(servicesToDelete.size(), applicationConfiguration.getServiceHandlingMaxParallelThreads());
+
+        return ForkJoinPoolUtil.execute(maxParallelThreads, () ->  getExistingServices(client, servicesToDelete));
+    }
+
+    private Set<CloudServiceInstanceExtended> getExistingServices(CloudControllerClient client, List<String> servicesToDelete) {
+        return servicesToDelete.parallelStream()
+                               .map(service -> getExistingService(client, buildCloudServiceExtended(service)))
+                               .filter(Objects::nonNull)
+                               .collect(Collectors.toSet());
+    }
+
+    @Override
+    protected Map<CloudServiceInstanceExtended, ServiceOperation>
+            getServicesInProgressState(ProcessContext context, Set<CloudServiceInstanceExtended> existingServices) {
+        CloudControllerClient client = context.getControllerClient();
+        int maxParallelThreads = Math.min(existingServices.size(), applicationConfiguration.getServiceHandlingMaxParallelThreads());
+
+        return ForkJoinPoolUtil.execute(maxParallelThreads, () -> getServicesInProgressStateInternal(client, existingServices));
+    }
+
+    private Map<CloudServiceInstanceExtended, ServiceOperation>
+            getServicesInProgressStateInternal(CloudControllerClient client, Set<CloudServiceInstanceExtended> existingServices) {
+        return existingServices.parallelStream()
```

Comment:
I am not really sure this could happen because a map would require "collecting" the parallel stream. However, we need the `ServiceOperation` object to be be calculated in the stream in order to take advantage of the parallelism.

Reply:
It will be calculated in the stream
The `collect` will be:
`existingServices.parallelStream().collect(service -> service, service -> serviceOpGetter.getLastServiceOp(client, service)`